### ISSUE TITLE
Chapter 3.4 & 3.5: Declartive vs Imperative

### ISSUE THREAD

**Issue by bbros-dev** [LukeMathWalker/zero-to-production#60]
In 3.5:

> Our spec for the health check endpoint was:
> > When we receive a GET request for /health_check we return a 200 OK response with no body.

I actually couldn't recall seeing that spec.  Might be useful to link to the relevant section?

But it raises the issue of writing specifications that are declarative rather than imperative.

In addition, for an integration spec there is a whole lot of implementation detail, and with:

> The test also covers the full range of properties we are interested to check:
>
> - the health check is exposed at /health_check
> - the health check is behind a GET method
> - the health check always returns a 200
> - the health check’s response has no body

I think you'll agree that in most of that spec you are actually testing `actix_web` rather than your app.
These, IMO, usually belong in unit tests, where the implementation detail can/does matter, sometimes its even been necessary for a unit to test validate the version of a library - I know the InSpec people say that should be a system level detail akin to their tests for configured SSH protocol version.
In regulated environments there can be a need to test that a specific method is used or not used in the implementation - but these sit most comfortably in the unit test suite.

I think you're initial context description (ch3, p18) had it mostly right:

> As a blog visitor,
> I want to subscribe to the newsletter,
> So that I can receive email updates when new content is published ~on the blog~.

Perhaps something similar can be added to ch 3.3 as the health check context:

> As a blog administrator,
> I want to be alerted when the sign-up service is unhealthy,
> So that I can restore the sign-up service

With a specification for alerts:

```gherkin
Given the sign-up service is monitored
When the sign-up service is unhealthy
Then the administrator receives an alert
```
 and a specification for no alerts:

```gherkin
Given the sign-up service is monitored
When the sign-up service is healthy
Then the administrator receives no alert
```

This way you don't change your integration specification files if you change your definition of "healthy" from 200 to 201 or 202 - contrived examples but not implausible. Of course your unit tests would change in that scenario.

Not sure if that helps?

**Comment by bbros-dev** [comment]
Current code-test workflow in Chapter 3:

1. Spec line
2. Function skeleton
3. Iterate function content
4. Integrate with App.new
5. Build
6. Retro fit test

An alternative for consideration:

1. Write specifcation for scenario:

   ```gherkin
	Feature: Newsletter sign-up service: Unhealthy status alerts
	  In order to restore the newsletter sign-up service
	  As a blog administrator,
	  I want to be alerted when the sign-up service is unhealthy

	  Background:
	    Given the sign-up service is monitored

	  Scenario: The sign-up service is unhealthy
	    When the sign-up service is unhealthy
	    Then the administrator receives one alert

	  Scenario: The sign-up service is healthy
	    When the sign-up service is healthy
	    Then the administrator receives no alert

   ```

2. Execute specification: No errors just counts of skipped unimplemented steps.:
   ```bash
     Finished test [unoptimized + debuginfo] target(s) in 0.28s
     Running target/debug/deps/zero2prod-336b234b8ab8c293

     running 0 tests

     test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in
     0.00s

     Running target/debug/deps/cucumber-c3684fce628ebd3d
     [Cucumber v0.8.4]
     Feature: Newsletter sign-up service: Unhealthy status alerts   features/health_check.feature:1:1 

      Scenario: The sign-up service is unhealthy           features/health_check.feature:9:3 
       - ⛓️ Given the sign-up service is monitored         features/health_check.feature:7:5 
         ⚡ Not yet implemented (skipped)
      Scenario: The sign-up service is healthy            features/health_check.feature:13:3 
       - ⛓️ Given the sign-up service is monitored         features/health_check.feature:7:5 
         ⚡ Not yet implemented (skipped)

     [Summary]
     1 features
     2 scenarios (2 skipped, 0 passed)
     4 steps (2 skipped, 0 passed)  
   ```
3. Write the missing Given step: This would either setup or confirm the monitoring service is activated. Right now it is a psuedo-dummy function that does somthing to show where you'd put code.

    ```rust
    #[given("the sign-up service is monitored")]
    async fn activate_monitor(world: &mut MyWorld) {
        world.foo = "monitor".into();
        world.test_async_fn().await;
    }
    ```

4. Execute the specification: The `Given` step is now green and we move on to the `When` step

    ```bash
	$ cargo test
        ....
        Feature: Newsletter sign-up service: Unhealthy status alerts
		                                              features/health_check.feature:1:1 

	 Scenario: The sign-up service is unhealthy           features/health_check.feature:9:3 
	  ✔ ⛓️ Given the sign-up service is monitored         features/health_check.feature:7:5 
	  - When the sign-up service is unhealthy            features/health_check.feature:10:5 
	    ⚡ Not yet implemented (skipped)
	 Scenario: The sign-up service is healthy            features/health_check.feature:13:3 
	  ✔ ⛓️ Given the sign-up service is monitored         features/health_check.feature:7:5 
	  - When the sign-up service is healthy              features/health_check.feature:14:5 
	    ⚡ Not yet implemented (skipped)

	[Summary]
	1 features
	2 scenarios (2 skipped, 0 passed)
	8 steps (2 skipped, 2 passed)

	Finished in 0.5 seconds.
    ```

5. Repeat Step 3:  Write the code you wish you had.  Then write a unit test. Check the unit test fails. **Now** write some code that makes the unit test pass.

6. Repeat Step 4
7. Rinse and repeat.

**Comment by bbros-dev** [comment]
@LukeMathWalker has addressed this in issue #58.

Closing as a off-topic/out of scope, since this is a topic on which the book is opinionated about - as advertised.
